src/bit-fields-macros/src/utils.rs

```diff
@@ -0,0 +1,228 @@
+/// small point, but helps a little.
+#[derive(Debug, Clone, Copy)]
+pub enum DataType {
+    /// `u8`
+    U8 = 8,
+    /// `u16`
+    U16 = 16,
+    /// `u32`
+    U32 = 32,
+    /// `u64`
+    U64 = 64,
+    /// `u128`
+    U128 = 128,
+}
+
+impl DataType {
+    // When const trait implementations are stabilized this should be
+    // `impl const From<DataType> for u8`.
+    /// Returns the size in bits of the data type.
+    #[allow(clippy::as_conversions)]
+    pub const fn size(self) -> u8 {
+        self as u8
+    }
+}
+
+impl fmt::Display for DataType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Self::U8 => write!(f, "u8"),
+            Self::U16 => write!(f, "u16"),
+            Self::U32 => write!(f, "u32"),
+            Self::U64 => write!(f, "u64"),
+            Self::U128 => write!(f, "u128"),
+        }
+    }
+}
+
+impl std::convert::TryFrom<&str> for DataType {
+    type Error = DataTypeTryFromError;
+    fn try_from(s: &str) -> Result<Self, Self::Error> {
+        match s {
+            "u8" => Ok(Self::U8),
+            "u16" => Ok(Self::U16),
+            "u32" => Ok(Self::U32),
+            "u64" => Ok(Self::U64),
+            "u128" => Ok(Self::U128),
+            _ => Err(DataTypeTryFromError),
+        }
+    }
+}
+
+/// Error type for `impl TryFrom<&str> for DataType`
+#[derive(Debug, PartialEq, Eq)]
+pub struct DataTypeTryFromError;
+
+impl fmt::Display for DataTypeTryFromError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "Bad backing data type given. Only accepted types are `u8`, `u16`, `u32`, `u64` and \
+             `u128`."
+        )
+    }
+}
+
+impl std::error::Error for DataTypeTryFromError {}
+
+/// Utility used to assembly the display string
+#[derive(Debug, Default)]
+pub struct MultiLineString(pub Vec<String>);
+
+impl MultiLineString {
+    /// Create new empty string.
+    pub fn new() -> Self {
+        Self(Vec::new())
+    }
+    /// Pushes `MultiLineString` on to `self`, if the given `other` is deeper than `self` it will
+    /// expand down to accommodate.
+    // None of these indexes or arithmetic operations will ever panic, thus we can safely use
+    // unchecked operations here. Since unchecked arithmetic is currently unstable we just allow
+    // integer arithmetic for now.
+    #[allow(clippy::integer_arithmetic, clippy::arithmetic)]
+    pub fn push(&mut self, other: &Self) {
+        // SAFETY:
+        // Always safe.
+        unsafe {
+            let mut i = 0;
+            while i < std::cmp::min(self.0.len(), other.0.len()) {
+                self.0
+                    .get_unchecked_mut(i)
+                    .push_str(other.0.get_unchecked(i));
+                i += 1;
+            }
+            while i < other.0.len() {
+                self.0.push(other.0.get_unchecked(i).clone());
+                i += 1;
+            }
+        }
```

Comment:
Perhaps a little over engineering, but I re-wrote it to use a functional approach:
```rust
pub fn push(&mut self, other: &Self) {
    // Split `other` into slice to join to `self` and slice to extend `self`.
    let (join,extend) = other.0.split_at(self.0.len());
    
    // Lines where both `self` and `other` have strings, join them
    let joined = self.0.iter().zip(join).map(|(x,y)|format!("{x}{y}"));

    // Lines `other` extends below `self` add them
    let chained = joined.chain(extend.iter().cloned());

    // Collect into new string
    self.0 = chained.collect();
}
```
Thing I'd like here is to do `self.0.into_iter()` instead of of `self.0.iter()`.

Does this look good to you?

Reply:
It's just bikeshedding at this point, but maybe keep the first line and replace the two iters with for-loops? So 

```rust
let (join, extend) = other.0.split_at(self.0.len());

for (base, ext) in self.0.iter_mut().join(join) {
  base.push_str(ext);
}

for new in extend {
  self.0.push(new.clone());
}
```

Reply:
I tend to prefer to use the fully functional apart rather than a combination. I think you can also loose some optimizations here when doing individual `push`s (not that this really matters in this case).